typing/typecore.ml

```diff
@@ -4682,12 +4683,12 @@ and type_cases ?in_function env ty_arg ty_res partial_flag loc caselist =
   let ty_arg_check =
     if do_init then
       (* Hack: use for_saving to copy variables too *)
-      Subst.type_expr (Subst.for_saving Subst.identity) ty_arg
-    else ty_arg
+      Subst.type_expr (Subst.for_saving Subst.identity) ty_arg'
+    else ty_arg'
   in
   let partial =
     if partial_flag then
-      check_partial ~lev env ty_arg_check loc cases
+      check_partial ~lev env (instance env ty_arg_check) loc cases
```

Comment:
Due to the `end_def` at line 4611, we are no longer at the same level as when the patterns were typed.
I suggest keeping `ty_arg_check` rather than an instance of it. This is perfectly fine from the point of view of levels.

Reply:
I agree with the first part.
I did the change only because it seemed to me that, except here, we never pass an expected type which might be general to `type_pat`. But this might be purely incidental.

Can you comment on that point before I revert the change as per your suggestion?

Reply:
I'm not sure this is the only case, but you are right that `expected_ty` is not handled as a type scheme in `type_pat`. I do not see any problem with doing that here, as the type is a fresh copy, and by unification its levels will be lowered. It might be cleaner to take an instance at an higher level.

Reply:
When you say "*by unification its levels will be lowered*" I think you mean that they will be lowered to the current level (that is, the one at which `check_partial` is called). 
Wouldn't that be equivalent to taking an instance at that same level before calling `check_partial`?

> It might be cleaner to take an instance at an higher level.

Assuming that "higher" means "deeper", I don't understand why you'd want to do that in this case.

Reply:
Strictly speaking, this is the levels in the type it is unified to.
But looking more carefully, I think you are right: at this point `type_pat` does not contain nested calls to `begin_def`, so the levels in the generated type will indeed be the level at which `type_pat` is called. So taking an instance should be OK.